### ISSUE TITLE
[Feature] Router GEMM should keep fp32 output under deterministic inference (DeepSeek V3/V4)

### ISSUE THREAD

**Issue by b8zhong** [sgl-project/sglang#34758]
### Checklist

- [x] If this is not a feature request but a general question, please start a discussion at https://github.com/sgl-project/sglang/discussions. Otherwise, it will be closed.
- [x] Please use English. Otherwise, it will be closed.

### Motivation

The DeepSeek router GEMM should output fp32 under deterministic inference too. Today it falls back to bf16, and that is not needed for batch invariance.

`python/sglang/srt/models/deepseek_v2.py:512`:

```python
if get_exec().deterministic.enable_deterministic_inference:
    return F.linear(hidden_states, self.weight, None)   # -> bf16 logits
```

Every other branch in the same function returns fp32: `dsv3_router_gemm(out_dtype=torch.float32)`, `linear_bf16_fp32`, and the prefill-CP path.

**fp32 output does not break batch invariance.** Batch invariance is about reduction order, not output dtype. Deterministic `F.linear` runs `matmul_kernel_persistent`, which accumulates in fp32 and already has an fp32 store branch (`batch_invariant_ops.py:167`) that is never reached. It is the same kernel with the same reduction; only the final store differs. The branch is unreachable because `_matmul_persistent_triton` allocates output as the input dtype (`batch_invariant_ops.py:185`). Same for the JIT `dsv3_router_gemm` kernel, where `kOutFloat` only changes `cast<OutT>(final_sum)`.

**Impact.** Router logits feed `noaux_tc` top-k over 256/384 experts with an fp32 correction bias. bf16 has 8 mantissa bits, so adjacent expert scores often sit within one ULP and rounding can change which experts are selected — a different answer, not a small numeric difference.

**Related bug.** `_mm_dtype_compat` (`batch_invariant_ops.py:967`), registered for `aten::mm.dtype`, does `matmul_persistent(...).to(out_dtype)`, i.e. fp32 -> bf16 -> fp32. Callers asking for fp32 get bf16 values.

### Proposed change

1. Add `out_dtype` (or a preallocated `output`) to `matmul_persistent` / `_matmul_persistent_triton` so the existing fp32 store branch is reachable. Tiling and accumulation order are unchanged.
2. Use it in `_mm_dtype_compat` instead of `.to(out_dtype)`.
3. In `DeepseekV2MoEGate.forward`, request fp32 out from the batch-invariant matmul instead of returning bare `F.linear`.
4. Keep one kernel for all M under deterministic mode — the normal path switches kernels at `max_router_gemm_tokens` (4 on SM100/103, else 16), and a token-count-dependent switch is batch-variant on its own.
5. `_matmul_persistent_deepgemm` allocates `out` as `a.dtype` for `deep_gemm.bf16_gemm_nn`; either find an fp32-out entry point or route fp32 requests to Triton.

Cost should be negligible — the GEMM is `[M, 7168] x [7168, 256]` and this only changes the store dtype.

### Related resources

- `python/sglang/srt/models/deepseek_v2.py:512`
- `python/sglang/srt/batch_invariant_ops/batch_invariant_ops.py:167,185,275,967`
- `python/sglang/kernels/jit/csrc/gemm/dsv3_router_gemm.cuh` — `kOutFloat`
- #12095 — added the `F.linear` fallback

**Comment by pol-resina** [comment]
can i work on that?
